Add support for LC_LINKER_OPTION

milend opened the pull request:
Adds support for reading `LC_LINKER_OPTION` load commands which can be founder in object files. The PR is split into two commits, the first adds the required C functions and the second one adds the `MKLoadCommand` subclass.

milend committed:
Add functions for LC_LINKER_OPTION load commands

milend committed:
Add MKLCLinkerOption class to handle LC_LINKER_OPTION commands

DeVaukz commented on libMachO/MachO_ABI/LoadCommands/load_command_linker_option.h:
```diff
@@ -0,0 +1,59 @@
+//! Iterate over the strings using a block.
+_mk_export void
+mk_load_command_linker_option_enumerate_strings(mk_load_command_ref load_command,
+                                                void (^enumerator)(const char* string, uint32_t index, char* stop));
```
libMachO can use modern C's `bool` type.  The `stop` parameter should be typed `bool*`.

DeVaukz commented on libMachO/MachO_ABI/LoadCommands/load_command_linker_option.c:
```diff
@@ -0,0 +1,167 @@
+static const char*
+mk_load_command_linker_option_get_string(mk_load_command_ref load_command, uint32_t requested_index, size_t *string_length) {
+    _MK_LOAD_COMMAND_NOT_NULL(load_command, return NULL);
+    _MK_LOAD_COMMAND_IS_A(load_command, _mk_load_command_linker_option_class, return NULL);
```
`mk_load_command_linker_option_get_string` is a private function (not in the header).  The null+type checks only need to appear in the API entry points (`mk_load_command_linker_option_enumerate_strings`).

DeVaukz commented on libMachO/MachO_ABI/LoadCommands/load_command_linker_option.c:
```diff
@@ -0,0 +1,167 @@
+        return current_string_len;
+    }
+    
+    return 0;
```
This API should permit the caller to pass a `NULL` `output`, in which case the function returns the buffer size required to hold the requested string (`current_string_len`).

DeVaukz commented on MachOKit/Load_Commands/MKLCLinkerOption.m:
```diff
@@ -0,0 +1,122 @@
+                NSError *stringError = nil;
+                MKCString *string = [[MKCString alloc] initWithOffset:offset fromParent:self error:&stringError];
+                if (string == nil) {
+                    MK_PUSH_UNDERLYING_WARNING(strings, stringError, @"Failed to instantiate build tool version at index " PRIi32 "", (self.nstrings - stringCount));
```
Please change the warning message.

DeVaukz commented:
Thank you for the PR

milend committed:
Use bool type instead of char for LC_LINKER_OPTION enumeration block

milend committed:
Consistently place pointer asterisk for LC_LINKER_OPTION enumeration function

milend committed:
Remove parameter checks in private function for LC_LINKER_OPTION

milend committed:
Fix up error message for LC_LINKER_OPTION

milend committed:
Adjust mk_load_command_linker_option_copy_string semantics

The function now copies partial strings and handles empty output
buffers (NULL or zero length).

milend commented:
@DeVaukz: Thanks for the speedy PR review! I've now addressed your feedback, see the new commits. Let me know if you have any further comments/questions 🙂

DeVaukz commented on libMachO/MachO_ABI/LoadCommands/load_command_linker_option.c:
```diff
@@ -0,0 +1,166 @@
+    size_t current_string_len = 0;
+    const char *current_option_start = mk_load_command_linker_option_get_string(load_command, requested_index, &current_string_len);
+    
+    if ((output == NULL) || (current_option_start != NULL && (current_string_len + 1 /* NULL char */) <= output_len)) {
```
This still does not match the semantics of the other APIs that copy out strings.  See then end of `_mk_mach_lc_str_copy` (called by `mk_load_command_rpath_copy_path`) for an example.

If the size of `output_len` is < `current_string_len`, then `output_len - 1` bytes should be copied, and the last byte set to the `NULL` terminator`.

milend replied:
@DeVaukz: No problem. The exact desired semantics were not clear to me beforehand, I've now adjusted them to match as requested (note: I've just updated the last commit in the PR).

DeVaukz commented:
Thank you!
